File: src/main/java/cses/range_queries/DynamicRangeSum.java
package cses.range_queries;

import java.io.*;
import java.util.*;

public class DynamicRangeSum {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int q = Integer.parseInt(st.nextToken());

        long[] numbers = new long[n + 1];
        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= n; i++) {
            numbers[i] = Long.parseLong(st.nextToken());
        }

        FenwickTree fenwick = new FenwickTree(n);
        for (int i = 1; i <= n; i++) {
            fenwick.update(i, numbers[i]);
        }

        for (int i = 0; i < q; i++) {
            st = new StringTokenizer(br.readLine());
            int type = Integer.parseInt(st.nextToken());

            if (type == 1) {
                int idx = Integer.parseInt(st.nextToken());
                long newVal = Long.parseLong(st.nextToken());
                long delta = newVal - numbers[idx];
                numbers[idx] = newVal;
                fenwick.update(idx, delta);
            } else {
                int left = Integer.parseInt(st.nextToken());
                int right = Integer.parseInt(st.nextToken());
                out.println(fenwick.rangeQuery(left, right));
            }
        }

        out.flush();
    }

    static class FenwickTree {
        long[] tree;

        FenwickTree(int size) {
            tree = new long[size + 1];
        }

        void update(int i, long delta) {
            while (i < tree.length) {
                tree[i] += delta;
                i += i & (-i);
            }
        }

        long query(int i) {
            long sum = 0;
            while (i > 0) {
                sum += tree[i];
                i -= i & (-i);
            }
            return sum;
        }

        long rangeQuery(int left, int right) {
            return query(right) - query(left - 1);
        }
    }
}

File: src/main/kotlin/cses/range_queries/DynamicRangeSum.kt
package cses.range_queries

fun main() {
//    val (n, q) = readln().split(" ").map { it.toInt() }
//    val numbers = readln().split(" ").map { it.toLong() }.toMutableList()
//
//    val queries: MutableList<List<Int>> = mutableListOf()
//    for (i in 1..q) {
//        queries.add(readln().split(" ").map { it.toInt() })
//    }

    val n = 8
    val q = 4
    val numbers = mutableListOf(3L, 2L, 4L, 5L, 1L, 1L, 5L, 3L)

    val queries = listOf(
        listOf(2, 1, 4),
        listOf(2, 5, 6),
        listOf(1, 3, 1),
        listOf(2, 1, 4)
    )

    val answer = solutionFenwick(n, numbers, queries)

    answer.forEach { println(it) }
}

/**
 * This worked in the static approach, init O(N) and query O(1)
 * but updates take O(N) which can be improved
 */
fun solution(n: Int, numbers: List<Int>, queries: List<List<Int>>): List<Int> {
    val prefix = IntArray(n+1)
    val result = mutableListOf<Int>()

    numbers.forEachIndexed { i, number ->
        prefix[i+1] = prefix[i] + number
    }

    queries.forEach { query ->
        when (query.first()) {
            1 -> {
                val index = query[1]
                val from = numbers[index-1]
                val to = query[2]
                val difference = from - to
                for (i in index..n) {
                    prefix[i] = prefix[i] - difference
                }
            }
            2 -> {
                val lower = query[1] - 1
                val upper = query[2]
                result.add(prefix[upper] - prefix[lower])
            }
        }
    }

    return result
}

/**
 * Fenwick tree, init O(NlogN), query O(logN), update O(logN)
 */
fun solutionFenwick(n: Int, numbers: MutableList<Long>, queries: List<List<Int>>): List<Long> {
    val fenwick = FenwickTree(n)
    val result = mutableListOf<Long>()

    numbers.forEachIndexed { i, number ->
        fenwick.update(i + 1, number)
    }

    queries.forEach { query ->
        when (query.first()) {
            1 -> {
                val index = query[1]
                val update = query[2].toLong()
                val delta = update - numbers[index - 1]
                numbers[index - 1] = update
                fenwick.update(index, delta)
            }
            2 -> {
                result.add(fenwick.rangeQuery(query[1], query[2]))
            }
        }
    }

    return result
}

// --- Fenwick Tree Implementation ---
class FenwickTree(size: Int) {
    private val tree = LongArray(size + 1)

    fun update(i: Int, delta: Long) {
        var index = i
        while (index < tree.size) {
            tree[index] += delta
            index += index and -index
        }
    }

    fun query(i: Int): Long {
        var index = i
        var sum = 0L
        while (index > 0) {
            sum += tree[index]
            index -= index and -index
        }
        return sum
    }

    fun rangeQuery(lower: Int, upper: Int): Long = query(upper) - query(lower - 1)
}
